### PR TITLE
Promote check constraint to not null on PG >= 12

### PR DESCRIPTION
**Pull request by solebared**
Addresses #34 .

### References
* Mailing list discussion: [Feature idea: use index when checking for NULLs before SET NOT NULL](https://www.postgresql.org/message-id/flat/7fc87d44-82de-4592-9cca-14536af274c3%40www.fastmail.com).
* [Commit](https://github.com/postgres/postgres/commit/bbb96c3704c041d139181c6601e5bc770e045d26) introducing the feature to PostgreSQL.
* [SO answer](https://dba.stackexchange.com/a/268128) from author of commit above.
* [Recipe](https://gist.github.com/jjb/fab5cc5f0e1b23af28694db4fc01c55a), with links to further reading

### Note on author's minimal expertise :)
I'm not really qualified to evaluate any risks associated with this approach. I've read through the references above and done some minimal testing locally, but don't have much personal expertise in PostgreSQL.

As such, i'm relying on reviewers to weigh in on the approach and potential gotchas.

### CODE REVIEW

File: lib/online_migrations/change_column_type_helpers.rb [hunk 1]
```diff
@@ -247,22 +248,7 @@ def finalize_columns_type_change(table_name, *column_names)
 
           # At this point we are sure there are no NULLs in this column
           transaction do
-            # For PG < 11 and Primary Key conversions, setting a column as the PK
-            # converts even check constraints to NOT NULL column constraints
-            # and forces an inline re-verification of the whole table.
```
Comment (solebared) [inline lib/online_migrations/change_column_type_helpers.rb:252]: The comments above seemed to belong closer to the '< 11' && PK codepath in `initialize_columns_type_change` so moved there.

The rest of the comments below were moved into the new `__set_not_null` private method.

File: test/schema_statements/changing_column_type_test.rb [hunk 2]
```diff
@@ -260,6 +260,22 @@ def test_finalize_column_type_change_copies_check_constraints
...
 
+    def test_finalize_column_type_change_preserves_not_null_without_default_before_12
+      with_postgres(11) do
+        @connection.initialize_column_type_change(:projects, :description, :text)
+        @connection.finalize_column_type_change(:projects, :description)
+        refute column_for(:projects, :description).null
+      end
+    end
+
+    def test_finalize_column_type_change_preserves_not_null_without_default_after_12
```
Comment (solebared) [inline test/schema_statements/changing_column_type_test.rb:271]: I couldn't figure out how to verify that >= 12 doesn't update `pg_attribute`...

Tried to explicitly `revoke update on pg_catalog.pg_attribute on #{@connection.raw_connection.user`, but this didn't seem to take.

As such, the `_before_12` and `_after_12` specs are identical in this first draft, except for the version specification.

**Comment by solebared** [comment]
Tested locally on v13.11, and am seeing the desired log output:
```
LOG:  statement: ALTER TABLE "some_numbers" VALIDATE CONSTRAINT "chk_rails_0171ff2f52" /*application:Wrapbook*/
DEBUG:  verifying table "some_numbers"
LOG:  statement: SET statement_timeout TO '0' /*application:Wrapbook*/
LOG:  statement: BEGIN /*application:Wrapbook*/
LOG:  statement: ALTER TABLE "some_numbers"
        ALTER "number_for_type_change"
        SET NOT NULL
         /*application:Wrapbook*/
DEBUG:  existing constraints on column "some_numbers.number_for_type_change" are sufficient to prove that it does not contain nulls
LOG:  execute <unnamed>: SELECT conname, pg_get_constraintdef(c.oid, true) AS constraintdef, c.convalidated AS valid
                    FROM pg_constraint c
                    JOIN pg_class t ON c.conrelid = t.oid
                    JOIN pg_namespace n ON n.oid = c.connamespace
                    WHERE c.contype = 'c'
                      AND t.relname = 'some_numbers'
                      AND n.nspname = ANY (current_schemas(false))
         /*application:Wrapbook*/
LOG:  statement: ALTER TABLE "some_numbers" DROP CONSTRAINT "chk_rails_0171ff2f52" /*application:Wrapbook*/
LOG:  statement: COMMIT /*application:Wrapbook*/
```
Note in particular the second DEBUG entry.

**Comment by fatkodima** [comment]
Thanks for the PR! 💪 
I will review it today/tomorrow, make some inline changes (if necessary) and release a new version.

**Pull request merged**